plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Utils.hs

```diff
@@ -99,3 +101,63 @@ showSpan x = show (srcSpanStartLine x) <> ":" <> show (srcSpanStartCol x) <> "-"
 mkRange :: (Integral a1, Integral a2) => a1 -> a2 -> a2 -> Range
 mkRange startLine startCol len =
     Range (Position (fromIntegral startLine) (fromIntegral startCol)) (Position (fromIntegral startLine) (fromIntegral $ startCol + len))
+
+
+-- nameLength is in code points unit.
+-- while Range might not in code points unit.
+-- but we can still use it to get the length
+-- since it is only used to exclude some names
+-- currently, we only break `(ModuleA.b)` into `(ModuleA` and `.b)`
+splitModuleNameAndOccName :: VirtualFile -> Range -> Identifier -> [(Range,Identifier)]
+splitModuleNameAndOccName _ ran (Left m) = [(ran, Left m)]
+splitModuleNameAndOccName vf ran@(Range (Position startLine startColumn) (Position _endLine endColumn)) (Right name)
+    | nameLength name < fromIntegral (endColumn - startColumn), (Just text) <- getTextByCodePointRangeFromVfs vf ran =
```

Comment:
I'm still confused about why we can't use the text from the `OccName`? In particular, doesn't `occNameString . nameOccName` give you what you want? You even use that below!

Reply:
What I would kind of expect is something like:
- Check if the text is delimted by parens or backticks, remember this
- Match the name text against a regex to pull out the module prefix if there is one
- Compute the ranges based on the previous things:
   - If delimited, start and end one character earlier
   - If there is a module prefix, add a token for that bit, and start the operator bit later

Reply:
I update the doc, should be clear now.

Reply:
> I'm still confused about why we can't use the text from the OccName? In particular, doesn't occNameString . nameOccName give you what you want? You even use that below!
the reason is here https://github.com/haskell/haskell-language-server/issues/3957#issuecomment-1893485590